expat/fuzz/xml_parse_fuzzer.c

```diff
@@ -47,18 +47,45 @@ end(void *userData, const XML_Char *name) {
   (void)name;
 }
 
-int
-LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
-  XML_Parser p = XML_ParserCreate(xstr(ENCODING_FOR_FUZZING));
-  assert(p);
-
+static void
+ParseOneInput(XML_Parser p, const uint8_t *data, size_t size) {
   // Set the hash salt using siphash to generate a deterministic hash.
   struct sipkey *key = sip_keyof(hash_key);
   XML_SetHashSalt(p, (unsigned long)siphash24(data, size, key));
 
   XML_SetElementHandler(p, start, end);
   XML_Parse(p, (const XML_Char *)data, size, 0);
-  XML_Parse(p, (const XML_Char *)data, size, 1);
+  if (XML_Parse(p, (const XML_Char *)data, size, 1) == XML_STATUS_ERROR) {
+      XML_ErrorString(XML_GetErrorCode(p));
+      XML_GetCurrentLineNumber(parser);
+  }
+  if (size % 2) {
+    XML_ParserReset(p, NULL);
+  }
+}
+
+int
+LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+  XML_Parser p = XML_ParserCreate(xstr(ENCODING_FOR_FUZZING));
+  assert(p);
+  ParseOneInput(p, data, size);
+  // not freed yet, but used later and freed then
+
+  XML_Parser pns = XML_ParserCreateNS(NULL, '!');
+  assert(pns);
+  ParseOneInput(pns, data, size);
+  XML_ParserFree(pns);
+
+  XML_Parser entParser = XML_ExternalEntityParserCreate(p, "e1", NULL);
+  assert(entParser);
+  ParseOneInput(entParser, data, size);
+  XML_ParserFree(entParser);
+
+  XML_Parser ext_parser = XML_ExternalEntityParserCreate(p, NULL, NULL);
+  assert(ext_parser);
+  ParseOneInput(ext_parser, data, size);
+  XML_ParserFree(ext_parser);
   XML_ParserFree(p);
+
```

Comment:
It would be nice to have the empty line before that special (parent parser) free, not after.

Reply:
ok